utils/filterPAS/src/filterPAS/functions.py

```diff
@@ -0,0 +1,197 @@
+    cat_dict = {col: lambda x: x.astype(str).str.cat(sep=sep)
+                  for col in df.columns
+                    if col not in distinct_cols}
+
+    # Cols want to report a single value (e.g. where expect same values in all rows/not important)
+    # Also important where need to maintain original
+    distinct_dict = {col: lambda x: x.first()
+                         for col in distinct_cols
+                        }
+
+    # Define the merged interval
+    cat_dict["Start"] = lambda x: x.min()
+    cat_dict["End"] = lambda x: x.max()
+
+    grpd = df.groupby(cluster_col, group_keys=False)
+
+    # Cols want to report a single value (e.g. where expect same values in all rows/ output value not important)
+    grpd_distinct = grpd[distinct_cols].first()
+
+    grpd_other = grpd[other_cols].agg(cat_dict)
+
+    # Combine to single row
+    clpsd = grpd_distinct.merge(grpd_other, left_index=True, right_index=True)
+
+    return clpsd[col_order]
+
+
+def collapse_tes(gr,
+                 cluster_col="Cluster",
+                 distinct_cols=["Feature", "gene_id"],
+                 sep=","):
+    '''
+    Collapse overlapping regions into union interval whilst preserving metadata
+    gr: pr.PyRanges object
+    cluster_col: Name of column grouping intervals wish to collapse. e.g. output of pr.cluster()
+    distinct_cols: list of column names wish to retain a single value for collapsed interval.
+        - All other columns in object will be collapsed to a single string
+        - Note: cluster_col, 'Chromosome' and 'Strand' will be added internally (required to produce a valid object)
+    sep: string character to separate values of all columns not found in distinct_cols
+    '''
+
+    # Want to keep grouping col to single value
+    # Chromosome and Strand must be a single value to output a valid PyRanges dataframe
+    dist_cols = distinct_cols + [cluster_col] + ["Chromosome", "Strand"]
+
+    return gr.apply(lambda df: _collapse_tes(df, cluster_col, dist_cols, sep=sep))
+
+
+def assign_id(gr, cols_to_cat=["Chromosome","Start","End","Strand"], sep_char=":", out_col="pas_id"):
+    '''
+    Generate a identifier column from a subset of column names
+    gr: pr.PyRanges object
+    cols_to_cat: list of columns names wish to combine (order of names defines order in output column)
+    sep_char: str to denote separator between values from columns defined in cols_to_cat
+    out_col: str denoting name of output column
+    '''
+
+    gr_cols = gr.columns.tolist()
+
+    for col in cols_to_cat:
+        assert col in gr_cols, f"Following column name - {col} - not found in input gr"
+
+    return gr.assign(out_col,
+                     lambda df: df[cols_to_cat[0]].str.cat(df[cols_to_cat[1:]].astype(str),
+                                                           sep=sep_char
+                                                           )
+                    )
```

Comment:
Is `functions.py` still required as have all the same functions stored in `main.py`?